server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java

```diff
@@ -1583,6 +1585,10 @@ private SearchSourceBuilder parseXContent(XContentParser parser, boolean checkTr
         return this;
     }
 
+    private void adjustKnnSizeParam() {
+        if (size > 0) knnSearch.forEach(knn -> knn.requestSize(size));
+    }
```

Comment:
It seems like this isn't being used at all within the KnnSearchBuilder. However, we can do better. 

We should have a `KnnSearchBuilder.Builder` that `KnnSearchBuilder.fromXContent` returns. That can have setters for `k` that gets adjusted after the last field is read when parsing the `SearchSourceBuilder`. 

This way we have a `KnnSearchBuilder.Builder#build(int)` where `int` is the request size (the default or user provided). And `KnnSearchBuilder` will never have `null` `k` nor null `num_candidates`.

Reply:
I see what you mean - Nice suggestion! 
Currently, `org.elasticsearch.search.builder.SearchSourceBuilder#parseXContent` doesn't do much outside of the `while` loop other than iterating through all objects. I guess we should add the call to the builder within this parsing section though, as only then the `SearchSourceBuilder` would have been finalized parsing. 

Do you think that we should hide all other public test usages/constructors of `KnnSearchBuilder` behind the builder (except from the one accepting a `StreamInput`), or just isolate this very specific use-case?

Reply:
> Do you think that we should hide all other public test usages/constructors of KnnSearchBuilder behind the builder (except from the one accepting a StreamInput), or just isolate this very specific use-case?

I think the ctor should be package private & we use a builder object for all `final` object values.

Reply:
>  And KnnSearchBuilder will never have null k nor null num_candidates.

So, since the responsibility to fill out missing values will be moved over to the `Builder` on the node that received the original request, we wouldn't need any transport versions or backwards compatibility as well, right? The values will have already been filled when sending over the objects, so the behavior right after `builder#build` will be pretty much the same.

Reply:
If the KnnSearchBuilder doesn't have null values, it won't need bwc serialization as nothing changes there. But, the KnnVectorQueryBuilder probably still needs it as `num_candidates` can be null when executed as knn-as-a-query

Reply:
Yeap, knn-query would still need it. But we can take it out of knn-search. Getting on it : )

Reply:
hey @pmpailis are you blocked by me here?

Reply:
Nope @benwtrent ! Was just taking a look at `toXContent` to understand under which circumstances this is invoked, and more specifically in the context of `knn-search`. Will have an update for that shortly :)